tests/test_webservice.py

```diff
@@ -219,16 +159,109 @@ def test_missing_constructor_args(self):
             Client("47")
 
 
-class TestFactors(BaseTest, unittest.TestCase):
+class BaseTransactionTest(BaseTest):
+    def has_ip_location(self):
+        return self.type in ["factors", "insights"]
+
+    def test_200(self):
+        model = self.create_success()
+        response = json.loads(self.response)
+        if self.has_ip_location():
+            response["ip_address"]["_locales"] = ("en",)
+        self.assertEqual(self.cls(response), model)
+        if self.has_ip_location():
+            self.assertEqual("United Kingdom", model.ip_address.country.name)
+
+    def test_200_on_request_with_nones(self):
+        model = self.create_success(
+            request={
+                "device": {"ip_address": "81.2.69.160", "accept_language": None},
+                "event": {"shop_id": None},
+                "shopping_cart": [{"category": None, "quantity": 2,}, None],
+            }
+        )
+        response = self.response
+        self.assertEqual(0.01, model.risk_score)
+
+    def test_200_with_locales(self):
+        locales = ("fr",)
+        client = Client(42, "abcdef123456", locales=locales)
+        model = self.create_success(client=client)
+        response = json.loads(self.response)
+        if self.has_ip_location():
+            response["ip_address"]["_locales"] = locales
+        self.assertEqual(self.cls(response), model)
+        if self.has_ip_location():
+            self.assertEqual("Royaume-Uni", model.ip_address.country.name)
+            self.assertEqual("Londres", model.ip_address.city.name)
+
+    def test_200_with_no_body(self):
+        with self.assertRaisesRegex(
+            MinFraudError,
+            "Received a 200 response but could not decode the response as"
+            " JSON: b?'?'?",
+        ):
+            self.create_success(text="")
+
+    def test_200_with_invalid_json(self):
+        with self.assertRaisesRegex(
+            MinFraudError,
+            "Received a 200 response but could not decode the response as"
+            " JSON: b?'?{'?",
+        ):
+            self.create_success(text="{")
+
+    def test_300_error(self):
```

Comment:
I didn't think that the Report Transactions API could return such a code. I guess anything's possible, and no harm is done by testing for it. I'll be happy to move it into the BaseTest class.

Reply:
Yeah, minFraud _shouldn't_ return this ether. Potentially it could happen with a Cloudflare misconfiguration or something. The test is more to see that the client does something sane if it did return something unexpected like this.